kong/cluster_events/init.lua

```diff
@@ -219,67 +277,27 @@ local function poll(self)
 
   min_at = min_at - self.poll_offset - 0.001
 
-  local max_at = ngx_now()
+  log(DEBUG, "polling events from: ", min_at)
 
-  log(DEBUG, "polling events from: ", min_at, " to: ", max_at)
-
-  for rows, err, page in self.strategy:select_interval(self.channels, min_at, max_at) do
+  for rows, err, page in self.strategy:select_interval(self.channels, min_at) do
     if err then
       return nil, "failed to retrieve events from DB: " .. err
     end
 
+    local count = #rows
+
     if page == 1 then
-      local ok, err = self.shm:safe_set(CURRENT_AT_KEY, max_at)
+      local at = count > 0 and rows[1].now or self.strategy:server_time() or ngx_now()
+      local ok, err = self.shm:safe_set(CURRENT_AT_KEY, at)
       if not ok then
         return nil, "failed to set 'at' in shm: " .. err
       end
     end
 
-    for i = 1, #rows do
-      local row = rows[i]
-
-      if row.node_id ~= self.node_id then
-        local ran, err = self.events_shm:get(row.id)
-        if err then
-          return nil, "failed to probe if event ran: " .. err
-        end
-
-        if not ran then
-          log(DEBUG, "new event (channel: '", row.channel, "') data: '",
-                     row.data, "' nbf: '", row.nbf or "none", "'")
-
-          local exptime = self.poll_interval + self.poll_offset
-
-          -- mark as ran before running in case of long-running callbacks
-          local ok, err = self.events_shm:set(row.id, true, exptime)
-          if not ok then
-            return nil, "failed to mark event as ran: " .. err
-          end
-
-          local cbs = self.callbacks[row.channel]
-          if cbs then
-            for j = 1, #cbs do
-              if not row.nbf then
-                -- unique callback run without delay
-                local ok, err = pcall(cbs[j], row.data)
-                if not ok and not ngx_debug then
-                  log(ERR, "callback threw an error: ", err)
-                end
-
-              else
-                -- unique callback run after some delay
-                local delay = max(row.nbf - ngx_now(), 0)
-
-                log(DEBUG, "delaying nbf event by ", delay, "s")
-
-                local ok, err = timer_at(delay, nbf_cb_handler, cbs[j], row)
-                if not ok then
-                  log(ERR, "failed to schedule nbf event timer: ", err)
-                end
-              end
-            end
-          end
-        end
+    for i = 1, count do
+      local ok, err = process_event(self, rows[i])
+      if not ok then
+        return nil, err
```

Comment:
I have one question to @thibaultcha about this. This didn't change (got only refactored), but I am a bit worried that could this lead to missing events because on line `291` we set `CURRENT_AT_KEY` to current time, which gets next time used as `min_at`. Should be rather log the error, or is the current ok?

Reply:
I mark this as resolved, we decided to move time on, only when there are new events.